src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java

```diff
@@ -0,0 +1,200 @@
+import bio.terra.cloudres.google.notebooks.AIPlatformNotebooksCow;
+import bio.terra.cloudres.google.notebooks.InstanceName;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.generated.model.ApiGcpAiNotebookInstanceContainerImage;
+import bio.terra.workspace.generated.model.ApiGcpAiNotebookInstanceCreationParameters;
+import bio.terra.workspace.generated.model.ApiGcpAiNotebookInstanceVmImage;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.ControlledAiNotebookInstanceResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.notebooks.v1.model.ContainerImage;
+import com.google.api.services.notebooks.v1.model.Instance;
+import com.google.api.services.notebooks.v1.model.Operation;
+import com.google.api.services.notebooks.v1.model.VmImage;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * A step for creating the AI Platform notebook instance in the Google cloud.
+ *
+ * <p>Undo deletes the created notebook instance.
+ */
+public class CreateAiNotebookInstanceStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(CreateAiNotebookInstanceStep.class);
+  private final CrlService crlService;
+  private final ControlledAiNotebookInstanceResource resource;
+  private final WorkspaceService workspaceService;
+
+  public CreateAiNotebookInstanceStep(
+      CrlService crlService,
+      ControlledAiNotebookInstanceResource resource,
+      WorkspaceService workspaceService) {
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    String projectId = workspaceService.getRequiredGcpProject(resource.getWorkspaceId());
+    InstanceName instanceName = createInstanceName(projectId);
+    Instance instance = createInstanceModel(flightContext, projectId);
+
+    AIPlatformNotebooksCow notebooks = crlService.getAIPlatformNotebooksCow();
+    try {
+      OperationCow<Operation> creationOperation;
+      try {
+        creationOperation =
+            notebooks
+                .operations()
+                .operationCow(notebooks.instances().create(instanceName, instance).execute());
+      } catch (GoogleJsonResponseException e) {
+        // If the instance already exists, this step must have already run successfully. Otherwise
```

Comment:
How can you tell the difference between the instance already existing from this flight and the instance existing from a previous flight creating it?

Reply:
I was going to rely on Jay's PR #291 to do a transaction on the ResourceDao to ensure uniqueness of the underlying cloud resource. As-is, it's broken as it will assume success.

I'll add a TODO to add a test to verify that multiple creations for the same instance fail once Jay's PR is merged.